msgspec/_json_schema.py

```diff
@@ -0,0 +1,381 @@
+                schema["anyOf"] = options
+            else:
+                schema.update(struct_schema)
+        elif len(structs) == 1:
+            _, subtype = structs.popitem()
+            options.append(_to_schema(subtype, name_map, ref_template))
+            schema["anyOf"] = options
+        else:
+            schema["anyOf"] = options
+    elif isinstance(t, mi.LiteralType):
+        schema["enum"] = sorted(t.values)
+    elif isinstance(t, mi.EnumType):
+        schema.setdefault("title", t.cls.__name__)
+        if _has_nondefault_docstring(t):
+            schema.setdefault("description", t.cls.__doc__)
+        schema["enum"] = sorted(e.value for e in t.cls)
+    elif isinstance(t, mi.StructType):
+        schema.setdefault("title", t.cls.__name__)
+        if _has_nondefault_docstring(t):
+            schema.setdefault("description", t.cls.__doc__)
+        required = []
+        names = []
+        fields = []
+
+        if t.tag_field is not None:
+            required.append(t.tag_field)
+            names.append(t.tag_field)
+            fields.append({"enum": [t.tag]})
+
+        for field in t.fields:
+            field_schema = _to_schema(field.type, name_map, ref_template)
+            if field.required:
+                required.append(field.encode_name)
+            elif field.default is not mi.UNSET:
+                field_schema["default"] = mi._roundtrip_json(field.default)
+            names.append(field.encode_name)
+            fields.append(field_schema)
+
+        if t.array_like:
+            n_trailing_defaults = 0
+            for n_trailing_defaults, f in enumerate(reversed(t.fields)):
+                if f.required:
+                    break
+            schema["type"] = "array"
+            schema["prefixItems"] = fields
+            schema["minItems"] = len(fields) - n_trailing_defaults
+            if t.forbid_unknown_fields:
+                schema["maxItems"] = len(fields)
+        else:
+            schema["type"] = "object"
+            schema["properties"] = dict(zip(names, fields))
+            schema["required"] = required
+            if t.forbid_unknown_fields:
+                schema["additionalProperties"] = False
+    elif isinstance(t, (mi.TypedDictType, mi.DataclassType, mi.NamedTupleType)):
+        schema.setdefault("title", t.cls.__name__)
+        if _has_nondefault_docstring(t):
+            schema.setdefault("description", t.cls.__doc__)
+        names = []
+        fields = []
+        required = []
+        for field in t.fields:
+            field_schema = _to_schema(field.type, name_map, ref_template)
+            if field.required:
+                required.append(field.encode_name)
+            elif field.default is not mi.UNSET:
+                field_schema["default"] = mi._roundtrip_json(field.default)
+            names.append(field.encode_name)
+            fields.append(field_schema)
+        if isinstance(t, mi.NamedTupleType):
+            schema["type"] = "array"
```

Comment:
Lots of repetition. I'd consider using enums or constants

Reply:
Can you expand on what you mean by that? How would enums or constants make this code more maintainable?